webTarget examples for JAX-RS Client feature config (#1059)
OpenLiberty/docs: ref/feature/jaxrsClient/examples.adoc

== Examples

> **LC: Could do with an introductory sentence here...**

Reviewer: ```suggestion
Here are some examples of how you can update JAX-RS Client behavior in the Open Liberty configuration.
```

=== Overriding JAX-RS Client configuration when moving to production
The JAX-RS Client APIs allow you to specify properties on the client that configure settings like timeouts, proxy
hosts, SSL properties, and so on. If the application is deployed in a different environment, these settings might need to change. For example, an application might initially be written to run in a development environment, where all remote RESTful services are available without a proxy server and can connect very quickly:

[source,java]
----
public Widget getRemoteWidget(String id) {
    Client client = ClientBuilder.newClient();
    WebTarget target = client.target("https://somehost:9443/WidgetStore/widgets/");
    target.property("com.ibm.ws.jaxrs.client.connection.timeout", 5000); // 5 seconds

>     target.property("myCustomProperty", "dev");

Reviewer: We should delete these lines.

    Widget widget = target.path(id).request().get(Widget.class);
    return widget;
}
----

> When this application is deployed in a production environment, however, it might be necessary to modify these settings. This can be accomplished by adding the following configuration to the server config (`server.xml`):

Reviewer: ```suggestion
When this application is deployed in a production environment, however, it might be necessary to add new settings. This can be accomplished by adding the following configuration to the server config (`server.xml`):
```

> **LC: I think this implies that you can override the Java values with the server.xml values when that isn't what happens. Can you change how this is explained? I assume you need to leave the values out of the Java code for this to work?**

Reviewer: Correct.  I've made suggestions to help clarify this.  Good catch!

[source,xml]
----
<webTarget uri="https://somehost:9443/WidgetStore*" proxyHost="myProxyHost" proxyPort="55555" proxyType="HTTP"

>            connectionTimeout="30000" receiveTimeout="30000" />

Reviewer: ```suggestion
           connectionTimeout="60000" receiveTimeout="60000" />
```

----

> This changed configuration alters the `WebTarget` in the Java code to use the specified HTTP proxy server (`myProxyHost:55555`) and use connection and receive timeouts of 30 seconds (30,000 milliseconds).

Reviewer: ```suggestion
This configuration alters the `WebTarget` in the Java code to use the specified HTTP proxy server (`myProxyHost:55555`) and use connection and receive timeouts of 60 seconds (60,000 milliseconds) instead of the default timeouts.
```

=== Simplifying SSL configuration in development

> During development, an SSL certificate might not match the server hostname on which the application is being tested because it matches the server hostname on which the application will ultimately be deployed in production. The configuration can be changed during development to use a different SSL configuration or to disable hostname verification of SSL certificates (this is not recommended in production environments). You can or update a custom property:

Reviewer: ```suggestion
During development, an SSL certificate might not match the server hostname on which the application is being tested because it matches the server hostname on which the application will ultimately be deployed in production. The configuration can be changed during development to use a different SSL configuration or to disable hostname verification of SSL certificates (this is not recommended in production environments):
```

[source,xml]
----
<webTarget uri="https://somehost:9443/WidgetStore*" sslConfig="mySSLRef" disableCNCheck="true" />
----

> **LC: Needs an example for why and how you'd use custom properties here, and maybe a sub-heading...**

Reviewer: ```suggestion
If your application uses providers such as a `ClientRequestFilter`, `ClientResponseFilter`, etc. then they could read and react to custom properties specified on the JAX-RS Client or WebTarget.  It is also possible to change Open Liberty's client implementation behavior by specifying custom properties.  These would normally be specified programmatically via `webTarget.property("myCustomProperty", "someValue")`.
```

> You can also specify custom properties for the `WebTarget`:

Reviewer: ```suggestion
You can also specify custom properties for the `WebTarget` in the configuration:
```
